### ISSUE TITLE
Remove space between number and percentage sign

### ISSUE THREAD

**Issue by mronkko** [StatTag/StatTagMac#23]
When I enter a field that is percentage formatted, StatTag puts a space between the number and percentage. E.g. "10 %". There should be either no space or this should be a configurable option. Both are used but no space is more common:

https://english.stackexchange.com/questions/3281/should-there-be-a-space-before-a-percent-sign

**Comment by lrasmus** [comment]
Thanks @mronkko!  The intent was for there **not** to be a space before the percent sign.  I'm considering this a bug instead of making it an option.

**Comment by ewhitley** [comment]
Hi, @lrasmus and @mronkko  - This is actually a function of the OS locale. 

EX: 
* if you set to US you don't get a space ("29.23%"). 
* If you set to "de" you do get a space ("29.23 %").

We intentionally honor the locale of the user to control all number formatting ("," vs. "." as separators, etc.).

This can be done, but the trade-off is it will force a single set of locale formats on that field, which could have adverse effects with decimal notation.

**Comment by ewhitley** [comment]
@lrasmus - you can see this in a sample Swift playground (ported from the Obj-C) below.

==========

```

import Cocoa


var currentLocale: NSLocale?
let DecimalPlaces = 2

func formatPercentage(_ value: String?) -> String? {
    let f = NumberFormatter()
    f.numberStyle = .decimal
    var currentLoc = NSLocale.current as NSLocale
    if currentLocale != nil {
        if let currentLocale = currentLocale {
            currentLoc = currentLocale
        }
    }
    f.locale = currentLoc as Locale

    let aNumber = f.number(from: value ?? "")

    let numericValue = aNumber?.doubleValue ?? 0.0

    let formatter = NumberFormatter()
    formatter.numberStyle = .percent
    formatter.minimumFractionDigits = DecimalPlaces
    formatter.maximumFractionDigits = DecimalPlaces
    formatter.roundingMode = .halfEven
    formatter.locale = currentLoc as Locale
    return formatter.string(from: NSNumber(value: numericValue))

}

currentLocale = (Locale(identifier:"de") as NSLocale)
let numberValue = ".29234"
let percentValue = formatPercentage(numberValue)
print(percentValue)
```

**Comment by mronkko** [comment]
Interesting. Yes, in that case it should not be changed. Thanks for the quick response.

My locale is FI and indeed in the Finnish language, there should be a space. I should probably change my work computer locale to US...

On 17. Dec 2020, at 16.14, ewhitley <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi, @lrasmus<https://github.com/lrasmus> and @mronkko<https://github.com/mronkko> - This is actually a function of the OS locale.

EX:

  *   if you set to US you don't get a space ("29.23%").
  *   If you set to "de" you do get a space ("29.23 %").

We intentionally honor the locale of the user to control all number formatting ("," vs. "." as separators, etc.).

This can be done, but the trade-off is it will force a single set of locale formats on that field, which could have adverse effects with decimal notation.

**Comment by ewhitley** [comment]
Microsoft also has equivalent guidelines on this https://docs.microsoft.com/en-us/globalization/locale/number-formatting which is controlled through regional formatting

```
The placement of the percent sign (%).
It can be written several ways: 98%, 98 %, 98 pct, %98. Thus you should never assume that you can hard-code the percent sign.
```

**Comment by ewhitley** [comment]
> Interesting. Yes, in that case it should not be changed. Thanks for the quick response. My locale is FI and indeed in the Finnish language, there should be a space. I should probably change my work computer locale to US... On 17. Dec 2020, at 

@mronkko We're 100% open to changing the behavior if you feel it would benefit from allowing variations. We can probably do something where we explicitly alter the % location post-formatting to address this in the UI. I just want to be careful about retaining the rest of the locale formatting defaults so we don't mangle the rest of the number. Ripping out the % and trimming the value to remove spaces would let us re-add any sort of custom % format you might want. No biggie. I just want to be sensitive to the locale portions of this.

**Comment by ewhitley** [comment]
@mronkko Wondering if I can get your feedback on something. Right now I'm looking at all aspects of number formatting as a function of the locale. If you se "en-us", for example, the thousands separator turns into "," and the percentage formatting is "(number)%" (no space). 

How do you see those things? Are they related? Do you see all aspects of number formatting as one overall "thing" that's set once or do you anticipate the ability to set each separately?

The reason I ask is that this feels like it should be an overall "per project" setting that's governed so that teams can collaborate and share StatTag documents while retaining one set of formatting rules. EX: I'm in the US and you're in Finland. We should be able to send the document back and forth and the number formatting should stay consistent between our workstations.

We elected to use the workstation locale to control formatting given the sheer number of locale-specific differences. I still believe that's appropriate, but - I do suspect it should be a project setting that one can change. For that to work (simply) the formatting for all numbers would need to be controlled (by default) by the project locale. Now - we could allow flexibility for that at a "type of field" level - but I wonder if a simple option is to be able to set the overall locale once for the project.

Just wondering how you see this.

**Comment by mronkko** [comment]
@ewhitley My first reaction was that these are related and should come from computer locale. 

But after thinking about it, I do not think that using computer locale is idea. The current way it works is likely to cause confusion. I consider myself quite technically competent, but I did not know that the space between the number and percent symbol is something that is determined by the locale. 

I am using a Mac and I set the locale to Finnish because I want all distances, currencies, temperatures, etc to be on the metric that I use. Some aspects can be customized and I have set the decimal separator to be "." and thousand separator to be " " instead of the default "," and "." for the locale. But not all aspects of locale, including the location of the percentage symbol can be customized.

My reference manager (Zotero) used to get the language setting from computer locale, but they moved it to be a per document setting (https://www.zotero.org/support/word_processor_plugin_usage#document_preferences). This is important for people who work in multilingual environments. For example, I have colleagues who write for both international (English language) journals and local journals (Finnish language).  In this case, you could allow the user to pick the locale for the document and have a short explanation below the setting that it affects all aspects of number formatting.

**Comment by ewhitley** [comment]
Do you think having the locale set at the document to be sufficient? If so - there's stuff (in the Mac version) to let us set document-level locale to override the system locale. We could default the document locale to the author's initial locale and the let that be changed. 

I'm not sure what the impact is to the Windows version with this, but I agree that setting locale on the document and then using that locale to drive all number, date, currency, etc. formatting would be a good approach. 

I just want to discover whether or not you'd want to then change individual aspects of the format for a single "type" of information (EX: percent symbol format placement, but not thousands separator).

**Comment by mronkko** [comment]
I do not think that changing individual aspects of the format is necessary. If a truly custom format is required, it is always possible to format the numbers in Stata or R and pass them to StatTag as strings.

**Comment by ewhitley** [comment]
Awesome. Ty. We'll discuss and see about best path forward.
